### ISSUE TITLE
Cookie decryption problem with nginx reverse proxy

### ISSUE THREAD

**Issue by mo-esmp** [VahidN/DNTCaptcha.Core#52]
I'm using DNTCaptcha with ASP.NET Core API application that is behind nginx reverse proxy. It seems DNTCaptcha cannot decrypt the cookie correctly and always returns BadRequest result.
Do you have any idea what can be wrong or give me any lead to check?

this is what I got from logs :
`isValidCookie:: 20 + 5�:���`

![req](https://user-images.githubusercontent.com/1659032/80942421-f54c1a80-8df9-11ea-87de-ce3b8dd36715.jpg)
![con-log](https://user-images.githubusercontent.com/1659032/80942430-fa10ce80-8df9-11ea-95f0-9be33d0c66c1.jpg)

## Environment
Windows Server
ASP.NET Core 3.1
Nginx 1.7.10
```
DNTCaptcha.Core version:  2.5.0
.NET Core SDK version:  3.1.201
IDE: 
```

**Comment by VahidN** [comment]
> each web farm node can't decrypt data that's encrypted by an app on any other node. The default configuration isn't generally appropriate for hosting apps in a web farm.
> You should .... always route user requests to the same node.
> 
> Consider a user who signs into the app using cookie authentication. The user signs into the app on one web farm node. If their next request arrives at the same node where they signed in, the app is able to decrypt the authentication cookie and allows access to the app's resource. If their next request arrives at a different node, the app can't decrypt the authentication cookie from the node where the user signed in, and authorization for the requested resource fails.
> 
> In short, you need to add the following code:
> services.AddDataProtection()
>     .SetApplicationName("shared app name")
>     .PersistKeysToFileSystem(new DirectoryInfo(@"\\server\share\directory\"));

More Info:
https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/web-farm?view=aspnetcore-3.1
https://stackoverflow.com/questions/56296644/asp-net-core-data-protection-on-web-farm

**Comment by VahidN** [comment]
[More info in Persian](https://www.dotnettips.info/post/2717/%d8%ba%db%8c%d8%b1%d9%85%d8%b9%d8%aa%d8%a8%d8%b1-%d8%b4%d8%af%d9%86-%da%a9%d9%88%da%a9%db%8c%e2%80%8c%d9%87%d8%a7%db%8c-%d8%a8%d8%b1%d9%86%d8%a7%d9%85%d9%87%e2%80%8c%d9%87%d8%a7%db%8c-asp-net-core-%d9%87%d8%a7%d8%b3%d8%aa-%d8%b4%d8%af%d9%87%e2%80%8c%db%8c-%d8%af%d8%b1-iis-%d9%be%d8%b3-%d8%a7%d8%b2-%d8%b1%db%8c%e2%80%8c%d8%a7%d8%b3%d8%aa%d8%a7%d8%b1%d8%aa-%d8%a2%d9%86#comment-17375)

**Comment by mo-esmp** [comment]
I have a single web server and the request will be delivered to the same server that generated the cookie. I will test the above guidance and let you know about the outcome. Thanks.

**Comment by thepra** [comment]
@mo-esmp Did you close the issue because you solved it?

**Comment by mo-esmp** [comment]
No, had no luck solving the problem.

**Comment by srestani** [comment]
I had the same problem in production environment. Analyzing the source code of DNTCaptcha.Core I found that it use remote IP as a salt in data encryption. The load balancer hides the original ip and DNTCaptcha.Core could not remove it. To solve I have configured the overriding from the X-Forwarder-For as explained here [Microsoft Docs](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-5.0#forwarded-headers-middleware-order) and I have added the IPs of the load balancer nodes as KnowNodes to exclude them. 
`            services.Configure<ForwardedHeadersOptions>(options =>`
`            {` 
`                 options.ForwardedHeaders = ForwardedHeaders.XForwardedFor | ForwardedHeaders.XForwardedProto;` 
`                 options.KnownProxies.Add(IPAddress.Parse("my load balancer ip 1"));` 
`                 options.KnownProxies.Add(IPAddress.Parse("my load balancer ip 2"));` 
`             });`

I hope it can be useful to someone.
